### ISSUE TITLE
PSP VFPU instructions `vpfxs` and family should not require brackets

### ISSUE THREAD

**Issue by bismurphy** [pspdev/binutils-gdb#9]
I am attempting to assemble some disassembled official PSP code. In the disassembly, I have a few instructions of `vpfxs` (as well as `vpfxt` and `vpfxd`). They have a syntax like:

`vpfxs      2, 1, 1, -2`

However, the `psp-as` in this repository fails to assemble this instruction, and says "Error: invalid operands `vpfxs 2,1,1,-2'"

Through some testing, I have determined that this is because the psp-as requires brackets around the operands. `vpfxs      [2, 1, 1, -2]` succeeds.

According to the official Sony docs at https://github.com/Decompollaborate/rabbitizer/files/11356335/VFPU-Instruction_Manual-English.pdf (page 272), this instruction should have 4 operands, but there is no bracketing around them. I believe the psp-as is requiring brackets that, according to the specification, should not exist.

I do not have enough familiarity with the instruction processing code to understand why these brackets are needed. It is possible that I am misunderstanding the documentation, but it doesn't seem like there should be any brackets in the operands. This is causing the binutils-gdb assembler to be incompatible with other tools (including existing disassemblers), and it would be great if it could follow the standard and be compatible. 

I did notice that the instruction definitions (here: https://github.com/pspdev/binutils-gdb/blob/1bf9b3f9be9d82cc89374ca916cd4e8e6115dcf8/opcodes/mips-opc.c#L3545) have a "k0" and "k1" listed; I'm not sure if those are creating the brackets. The definition syntax is a bit opaque.

An easy minimal example to use is:
`echo "vpfxs 2, 1, 1, -2" | pspdev/bin/psp-as -EL -march=allegrex -mabi=32 -Iinclude -o assembled_result.o -G0`
`echo "vpfxs [2, 1, 1, -2]" | pspdev/bin/psp-as -EL -march=allegrex -mabi=32 -Iinclude -o assembled_result.o -G0`

The first throws an operand error, the second builds successfully. If I am reading the documentation properly, this is the opposite of the proper behavior.

**Comment by bismurphy** [comment]
This repo doesn't have much activity, so I'll go ahead and ping @fjtrujy who seems to be the primary contributor :) Not trying to be too much of a bother, just don't want things to go unseen. Thanks!

**Comment by fjtrujy** [comment]
@davidgfnet is our man here!

**Comment by davidgfnet** [comment]
Hey there!
This is indeed true, during the last rewrite of the binutils patches, I had to do some compromises and this was one of them. Since most people don't use the prefix instructions directly this was never an issue, but it is indeed a breaking change.

I have to rebase these patches to the latest binutils, since I'm trying to upstream this work. Some small patches were already upstreamed. I will give it a shot and try to maintain the original syntax this time. Sorry for breaking you!

**Comment by bismurphy** [comment]
Understood, thanks for the update! Great to confirm that this is a known thing and that hopefully there is a path toward restoring the original syntax. While new homebrew devs indeed likely don't use these instructions, their use in retail games means having these kinds of tools will be super helpful as the reverse-engineering infrastructure for PSP matures. 

Please let me know if anything shifts on this - if it becomes a "won't fix" that's good to know, so that we can at least start looking for workarounds for the bracketed syntax. I have some ideas, but it would be great if this can end up working natively :)

Thanks for your help, best of luck with the rebasing!

**Comment by bismurphy** [comment]
Hi @davidgfnet , just following up on this to check up on whether patching this mismatch is something being looked into. Just hoping to make sure this doesn't end up being forgotten. Thanks!

**Comment by davidgfnet** [comment]
Yes, this is in my radar, just didnt have a lot of time lately. I hope this will be fixed in ~2025.
This requires a bit of rebasing (some already started in version 2.41, see https://lists.gnu.org/archive/html/info-gnu/2023-07/msg00009.html) and could take a bit. However we should have an internal version much sooner :)
Thank you!

**Comment by davidgfnet** [comment]
Heads up, I just finished rebasing to 2.44. I fixed this issue :)
We will test this a bit more and eventually use it in the main toolchain.
Some small changes could happen in the future, as I attempt upstreaming some of this code

**Comment by bismurphy** [comment]
Excellent news! Thank you for the notice.

**Comment by fjtrujy** [comment]
@bismurphy, the changes have been applied already. Please double-check it, reopen the issue if it is not working for you.
https://github.com/pspdev/psptoolchain-allegrex/pull/64

Cheers
